### ISSUE TITLE
runc "no space left on device" cgroup errors on concourse worker

### ISSUE THREAD

**Issue by JesseTAlford** [cloudfoundry/garden-runc-release#15]
## Description
When running Concourse tasks, one of our three workers is giving the following error:
```
runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused "mkdir /sys/fs/cgroup/memory/24d5d8e9-250f-4bc8-40ed-03a9d4f9ad52: no space left on device"
```

Here's an example of a build with the problem: https://runtime.ci.cf-app.com/teams/main/pipelines/cf-bosh-2-0/jobs/update-releases/builds/62

The vm in question is not especially full in any respect:
```
| VM                                              | State   | AZ | VM Type       | IPs         |         Load          | CPU  | CPU  | CPU  | Memory Usage | Swap Usage  | System     | Ephemeral  | Persistent |
|                                                 |         |    |               |             | (avg01, avg05, avg15) | User | Sys  | Wait |              |             | Disk Usage | Disk Usage | Disk Usage |
| worker/1 (56bd4970-30e6-4ae8-b4f2-546bb566561b) | running | z1 | workers       | 10.10.48.14 | 0.18, 0.46, 0.61      | 0.4% | 0.2% | 0.0% | 33% (9.8G)   | 1% (196.2M) | 59%        | 19%        | n/a        |
```

The only potentially "weird" configuration we've got on there is: 
```
garden:
  max_containers: 1000
  network_pool: "10.254.0.0/20"
```

But, we're not really using a lot of that, as `fly workers` reports:
```
56bd4970-30e6-4ae8-b4f2-546bb566561b  172         linux     none
```

Suraci said this isn't a disk issue, probably isn't a concourse issue, and suggested we give 'yall a shot at it.

### Setup Information
- Concourse versions: 1.6 & 2.0.2
- garden-runc versions: 0.4.0 & 0.6.0
- bosh stemcell: 3232.6

## Logging and/or test output
```
garden/garden.stdout.log
83381-{"timestamp":"1472766267.733082533","source":"guardian","message":"guardian.destroy.destroy.finished","log_level":1,"data":{"handle":"67234b85-750c-47fe-7b2e-921740c5a5b4","session":"998.2"}}
83382-{"timestamp":"1472766267.733111382","source":"guardian","message":"guardian.destroy.finished","log_level":1,"data":{"handle":"67234b85-750c-47fe-7b2e-921740c5a5b4","session":"998"}}
83383-{"timestamp":"1472766267.733154535","source":"guardian","message":"guardian.layer-already-deleted-skipping","log_level":1,"data":{"error":"could not find image: no such id: 37870bde9a1de1d445ed2d7862d2c17d75212ac33b1fe1810cc0401485013b8a","graphID":"67234b85-750c-47fe-7b2e-921740c5a5b4","id":"67234b85-750c-47fe-7b2e-921740c5a5b4"}}
83384-{"timestamp":"1472766267.733173132","source":"guardian","message":"guardian.create.create-failed-cleaningup.cleanedup","log_level":1,"data":{"cause":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/67234b85-750c-47fe-7b2e-921740c5a5b4: no space left on device\"","handle":"67234b85-750c-47fe-7b2e-921740c5a5b4","session":"997.3"}}
83385-{"timestamp":"1472766267.733195305","source":"guardian","message":"guardian.api.garden-server.create.failed","log_level":2,"data":{"error":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/67234b85-750c-47fe-7b2e-921740c5a5b4: no space left on device\"","request":{"Handle":"","GraceTime":0,"RootFSPath":"raw:///var/vcap/data/baggageclaim/volumes/live/25d4f36d-9f8f-4cff-4ac1-d441db8e8b71/volume","BindMounts":null,"Network":"","Privileged":true,"Limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{}}},"session":"3.1.517"}}
83386:{"timestamp":"1472766268.500758410","source":"guardian","message":"guardian.create.start","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999"}}
83387:{"timestamp":"1472766268.500808477","source":"guardian","message":"guardian.create.gc.start","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.1"}}
83388:{"timestamp":"1472766268.500828743","source":"guardian","message":"guardian.create.gc.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.1"}}
83389:{"timestamp":"1472766268.500849724","source":"guardian","message":"guardian.create.containerizer-create.start","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2"}}
83390:{"timestamp":"1472766268.521710157","source":"guardian","message":"guardian.create.containerizer-create.depot-create.started","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2.1"}}
83391:{"timestamp":"1472766268.521897554","source":"guardian","message":"guardian.create.containerizer-create.depot-create.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2.1"}}
83392:{"timestamp":"1472766268.521925688","source":"guardian","message":"guardian.create.containerizer-create.lookup.started","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2.2"}}
83393:{"timestamp":"1472766268.521943808","source":"guardian","message":"guardian.create.containerizer-create.lookup.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2.2"}}
83394:{"timestamp":"1472766268.521981955","source":"guardian","message":"guardian.create.containerizer-create.create.creating","log_level":1,"data":{"bundle":"/var/vcap/data/garden/depot/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","bundlePath":"/var/vcap/data/garden/depot/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","id":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","logPath":"/var/vcap/data/garden/depot/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5/create.log","pidFilePath":"/var/vcap/data/garden/depot/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5/pidfile","runc":"runc","session":"999.2.3"}}
83395:{"timestamp":"1472766268.564662218","source":"guardian","message":"guardian.create.containerizer-create.create.finished","log_level":1,"data":{"bundle":"/var/vcap/data/garden/depot/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2.3"}}
83396:{"timestamp":"1472766268.564709902","source":"guardian","message":"guardian.create.containerizer-create.runtime-create-failed","log_level":2,"data":{"error":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5: no space left on device\"","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2"}}
83397:{"timestamp":"1472766268.564728737","source":"guardian","message":"guardian.create.containerizer-create.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.2"}}
83398:{"timestamp":"1472766268.564753771","source":"guardian","message":"guardian.create.create-failed-cleaningup.start","log_level":1,"data":{"cause":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5: no space left on device\"","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.3"}}
83399:{"timestamp":"1472766268.564774513","source":"guardian","message":"guardian.destroy.started","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000"}}
83400:{"timestamp":"1472766268.564793587","source":"guardian","message":"guardian.destroy.state.started","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000.1"}}
83401:{"timestamp":"1472766268.572014570","source":"guardian","message":"guardian.destroy.state.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000.1"}}
83402:{"timestamp":"1472766268.572041273","source":"guardian","message":"guardian.destroy.state-failed-skipping-delete","log_level":1,"data":{"error":"runc state: runc create: exit status 1: open /run/runc/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5/state.json: no such file or directory","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000"}}
83403:{"timestamp":"1472766268.572063684","source":"guardian","message":"guardian.destroy.destroy.started","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000.2"}}
83404:{"timestamp":"1472766268.572134256","source":"guardian","message":"guardian.destroy.destroy.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000.2"}}
83405:{"timestamp":"1472766268.572152138","source":"guardian","message":"guardian.destroy.finished","log_level":1,"data":{"handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"1000"}}
83406:{"timestamp":"1472766268.572190762","source":"guardian","message":"guardian.layer-already-deleted-skipping","log_level":1,"data":{"error":"could not find image: no such id: dd7bb4a2c64e30b97b6bee40971fa86e1b134e2a30a940b22eeb9823be188611","graphID":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","id":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5"}}
83407:{"timestamp":"1472766268.572208643","source":"guardian","message":"guardian.create.create-failed-cleaningup.cleanedup","log_level":1,"data":{"cause":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5: no space left on device\"","handle":"d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5","session":"999.3"}}
83408:{"timestamp":"1472766268.572227478","source":"guardian","message":"guardian.api.garden-server.create.failed","log_level":2,"data":{"error":"runc create: exit status 1: process_linux.go:258: applying cgroup configuration for process caused \"mkdir /sys/fs/cgroup/memory/d7581e52-2e88-4cdb-5d99-2fa93b4b1cd5: no space left on device\"","request":{"Handle":"","GraceTime":0,"RootFSPath":"raw:///var/vcap/data/baggageclaim/volumes/live/29dd5ffc-c577-4b89-7b66-dd87f7cfd996/volume","BindMounts":[{"src_path":"/var/vcap/data/baggageclaim/volumes/live/a40b857f-8711-41a6-45a9-e82871eacfd4/volume","dst_path":"/tmp/build/get","mode":1}],"Network":"","Privileged":true,"Limits":{"bandwidth_limits":{},"cpu_limits":{},"disk_limits":{},"memory_limits":{}}},"session":"3.1.518"}}
83409-{"timestamp":"1472766268.948507786","source":"guardian","message":"guardian.list-containers.starting","log_level":1,"data":{"session":"1001"}}
83410-{"timestamp":"1472766268.949084759","source":"guardian","message":"guardian.list-containers.finished","log_level":1,"data":{"session":"1001"}}
83411-{"timestamp":"1472766280.467016459","source":"guardian","message":"guardian.create.start","log_level":1,"data":{"handle":"dfd74a82-61d6-4aa8-5771-3a01c729b534","session":"1002"}}
83412-{"timestamp":"1472766280.467072964","source":"guardian","message":"guardian.create.gc.start","log_level":1,"data":{"handle":"dfd74a82-61d6-4aa8-5771-3a01c729b534","session":"1002.1"}}
83413-{"timestamp":"1472766280.467092991","source":"guardian","message":"guardian.create.gc.finished","log_level":1,"data":{"handle":"dfd74a82-61d6-4aa8-5771-3a01c729b534","session":"1002.1"}}
```

## Accessing Our Broken Environment
We've used `monit stop beacon` to stop this worker from participating in Concourse, but preserve it for any investigation you might like to do.

Please feel free to jump on the environment and poke around. If you want to turn the beacon back on and re-run jobs, the jobs in the cf-release smokes display group or any of the jobs in [this](https://runtime.ci.cf-app.com/teams/main/pipelines/build-docker-images) pipeline should be fine.

We've also made read access to the repo containing our bosh manifest available to your team:
- www.github.com/cloudfoundry/runtime-ci-env

This is where the bosh manifest for concourse lives, in deployments/concourse.yml). The keypair for ssh'ing into bosh jobs is in `keypair`.

**Comment by garden-gnome** [comment]
Hi there!

We use Pivotal Tracker to provide visibility into what our team is working on. A story for this issue has been automatically created.

The current status is as follows:

* [ ] [#129667821](https://www.pivotaltracker.com/story/show/129667821) runc "no space left on device" cgroup errors on concourse worker


This comment, as well as the labels on the issue, will be automatically updated as the status in Tracker changes.

**Comment by teddyking** [comment]
Thanks for the clear description and info on how to access the env @JesseTAlford !
I'm having difficulty SSHing to the worker, is the SSH port for the BOSH director locked down to a specific set of IP addresses? I'm getting a `Errno::ETIMEDOUT` from here.

**Comment by JesseTAlford** [comment]
Yeah, it looks like there's a security group rule that restricts access; you hit the nail on the head, it only allows traffic from SF.

I've added a rule that allows traffic from `80.169.160.158/30`, which IOPS tells me is the Pivotal London office. If that doesn't work, you might try VPNing into the SF office, or letting me know what range you need a rule for.

Please let me know if that gets you in or not!

**Comment by teddyking** [comment]
Hi @JesseTAlford,

Thanks for sorting that out, we've now been able to gain access to the worker VM to take a look around.
We have some good news and some not so good news.
The good news is that we've found the cause of the `no space left on device` error you're seeing.
The reason for the error is that the maximum no. of memory Cgroups has been reached on the VM:

```
$ cat /proc/cgroups
#subsys_name    hierarchy       num_cgroups     enabled
cpuset  		1       		1524    		1
cpu     		2       		168     		1
cpuacct 		3       		168     		1
memory  		4       		**65535**  		1
```

Unfortunately we're not sure if this is something that can be fixed immediately...
Long story short is that we _believe_ we need [this feature](https://github.com/opencontainers/runc/issues/984) to be implemented in runc to prevent this from happening again.
It's looking like memory Cgroups are not getting cleaned up as efficiently as they could be, and so in environments that are creating lots and lots of containers, the aforementioned limit can be reached.

An interim solution is to upgrade the stemcell for the deployment, which, at the very least, will reset the count back down to 0.
It's also possible that the kernel in the newer stemcell will help with the mem Cgroup cleanup as well (but we're not 100% sure on that).

Thanks,
Ed & Petar

**Comment by garden-gnome** [comment]
Hello again!

All stories related to this issue have been accepted, so I'm going to automatically close this issue.

At the time of writing, the following stories have been accepted:

* [#129667821](https://www.pivotaltracker.com/story/show/129667821) **Investigate**: runc "no space left on device" cgroup errors on concourse worker


If you feel there is still more to be done, or if you have any questions, leave a comment and we'll reopen if necessary!

**Comment by JesseTAlford** [comment]
Cool, we used `bosh recreate worker 1` to resolve the problem. We'll just keep applying that balm until that runc-feature-based fix lands.
